### PR TITLE
Module3

### PR DESCRIPTION
**Pull request by I-Georg**
Please make sure you are raising this PR against your own repository and not the original. If it says https://github.com/CorndelWithSoftwire in the search bar right now, you are in the wrong place!

Please also don't forget to submit the PR's url to Aptem afterwards (by clicking on the corresponding exercise component, pasting the url in the textbox that appears and then pressing the finish button).

### CODE REVIEW

File: .env.test
Comment (JackMeadDev) [inline .env.test]: I tested and I don't think that these are valid credentials, but I'd recommend making this really obvious, (e.g. `TOKEN=this-is-a-test-token`) so that other developers don't have to think about whether this could be a credential leak, and to make it obvious to you if you ever accidentally updated these to genuine credentials

File: README.md
Comment (JackMeadDev) [inline README.md:61]: Worth noting that these are actually `TOID = {list_id}` etc. rather than the variable names you assign in the code

It would also be nice to specify the keys in the `.env.template` to help users who scan over this

File: README.md
Comment (JackMeadDev) [inline README.md]: This is one way of running the tests, but we should use poetry as we're using that for our package dependencies elsewhere. I.e. if you `poetry add pytest` (& `poetry add selenium` for the e2e tests), then future developers will pick those up by default when they run `poetry install`. They can then run the tests with `poetry run pytest <path_to_tests>`

File: README.md
Comment (JackMeadDev) [inline README.md]: Your e2e test should spin up the app without the user having to manually run the app separately

File: todo_app/.env.test.py
Comment (JackMeadDev) [inline todo_app/.env.test.py:1]: You shouldn't need to specify a python version of this file in addition to the `.env.test` file

File: todo_app/app.py
Comment (JackMeadDev) [inline todo_app/app.py]: Let's delete this commented code now you've got this working

File: todo_app/app.py
Comment (JackMeadDev) [inline todo_app/app.py]: Rather than use hardcoded list ids, we should use the environment variables, otherwise our tests can only ever hit our live board

File: todo_app/app.py
Comment (JackMeadDev) [inline todo_app/app.py]: (For simplicity, I think you can remove this board id & it won't change anything)

File: todo_app/integration_test.py
Comment (JackMeadDev) [inline todo_app/integration_test.py]: Here you're hitting trello directly, rather than hitting your application to test that its working - so this test isn't actually testing our application at all

File: todo_app/integration_test.py
Comment (JackMeadDev) [inline todo_app/integration_test.py]: You're mocking the lists endpoint, but you'll find you also need to mock the cards endpoint (`"https://api.trello.com/1/lists/{idList}/cards"`) since you also hit that one as part of your index function

File: todo_app/items_test.py
Comment (JackMeadDev) [inline todo_app/items_test.py]: If this is your unit test, we really want it to focus closely on a single unit, in this case the ViewModel class, whereas here you're testing a function that only exists for the test file.

An example of a unit test for the show_all_done_items might look like below

```python
def test_show_all_done_items_true_if_1_done_item():
	to_do_items = []
	doing_items = []
	done_items = [
		ToDo('id', 'name', date.today()),
	]
	view_model = ViewModel(to_do_items, doing_items, done_items)

	assert view_model.show_all_done_items() == True
```

This tight focus on a single unit is really important - we avoid the test from worrying about environment variables, communicating with trello etc.

File: todo_app/e2e_test.py
Comment (JackMeadDev) [inline todo_app/e2e_test.py]: We should make sure that these tests can be run against a fresh environment, rather than directly against your local environment that needs to be running at the same time, i.e. using this create_trello_board etc.

**Review by JackMeadDev — CHANGES_REQUESTED** [review]
This is good work with getting the new functionality working on your app, but the tests look confused; the unit/integration tests aren't quite checking what we'd expect, and the end-to-end tests look good but only run against your development board - if this was a real app this might mean we're creating todos against the live board every time we run our tests. I've left some thoughts below on those, but there's a lot going on in this module, so please do reach out via slack/email or indeed book a support call with any of the tutors if that would be helpful to discuss this in more detail

File: .env.test
Comment (I-Georg) [reply .env.test]: Thank you. I have updated them

File: README.md
Comment (I-Georg) [reply README.md:61]: Thank you. I have updated them

File: todo_app/integration_test.py
Comment (I-Georg) [reply todo_app/integration_test.py]: Updated, thank you

File: todo_app/items_test.py
Comment (I-Georg) [reply todo_app/items_test.py]: I have added the test, thank you

File: todo_app/app.py
Comment (JackMeadDev) [inline todo_app/app.py]: Let's delete old code rather than commenting

File: todo_app/items_test.py
Comment (JackMeadDev) [inline todo_app/items_test.py]: Let's delete all of these lines

File: todo_app/app.py
Comment (JackMeadDev) [inline todo_app/app.py]: Here and in several methods below you've hardcoded the list ids, even though you have them available from the environment variables (i.e. with toDoId, DoingId & DoneId) - this means that the code can only work with your existing board, and isn't fully configured which means the e2e test still isn't using the fresh board completely. We should adjust all of these, e.g. here to:
```
		query = {
	    	        'id': i,
			'key': os.environ['KEY'],
			'token' : os.environ['TOKEN'],
			'idList': doneId,
		}
```

File: todo_app/e2e_test.py
Comment (JackMeadDev) [inline todo_app/e2e_test.py]: The environment variable you check for in the code is `DOINGID` rather than `DOING`

File: todo_app/e2e_test.py
Comment (JackMeadDev) [inline todo_app/e2e_test.py]: The Firefox windows don't close after the tests because we haven't closed the driver, we should either do that manually with:
```python
    driver = webdriver.Firefox()
    yield driver
    driver.close()
```
or we can use a with block to handle that for us:
```python
    with webdriver.Firefox() as driver:
        yield driver
```

File: todo_app/flask_config.py
Comment (JackMeadDev) [inline todo_app/flask_config.py:4]: I don't think we're using this class any more, so we can remove it entirely now

File: todo_app/integration_test.py
Comment (JackMeadDev) [inline todo_app/integration_test.py]: We never use this board id so we can delete it here

File: todo_app/e2e_test.py
Comment (JackMeadDev) [inline todo_app/e2e_test.py]: These paths are very specific and make it both hard to check/adjust and also very fragile to changes.  Once we fix the fact it's not correctly overriding the environment variables, this doesn't find the right input. 
A much cleaner and more reliable xpath for this update to done might be more like `'//input[@type = "submit" and @value="Update to done"]'`

File: todo_app/e2e_test.py
Comment (JackMeadDev) [inline todo_app/e2e_test.py:85]: It would be nice to also make an assertion after this, to verify that the expected behaviour is actually happening when we click. For example, we could verify that a todo does now exist on the page with the option to send it back to the todo state;
```python
undo_completion_input = '//input[@type = "submit" and @value="Return to To do"]'
assert driver.find_element_by_xpath(undo_completion_input ) != None
```

File: README.md
Comment (JackMeadDev) [inline README.md]: Now that you're tracking this inside poetry, you shouldn't need to install either this or pytest with pip - that should be handled when the user runs a `poetry install`. This does mean that we're expecting the user to execute the tests with `poetry run pytest ...` rather than just `pytest` though

File: README.md
Comment (JackMeadDev) [reply README.md]: Don't forget to clear out this comment once it isn't the case

File: todo_app/.env.test.py
Comment (JackMeadDev) [inline todo_app/.env.test.py]: (Same comment on trying to avoid fake credentials that look real!)

File: todo_app/items_test.py
Comment (JackMeadDev) [reply todo_app/items_test.py]: These look much more like unit tests, so we can now get rid of this getItems function

File: todo_app/integration_test.py
Comment (JackMeadDev) [inline todo_app/integration_test.py:32]: We should make an assertion after this call to check that we're not getting any issues - just because the test hasn't errored doesn't mean it's succeeded!
E.g. here we could verify that some of your fake responses are being displayed on the page, a simple example might be `assert "test data" in response.data.decode()`

**Review by JackMeadDev — CHANGES_REQUESTED** [review]
Nice changes, your unit tests look much more focussed now, and the integration test is effectively mocking things. Your end-to-end test isn't quite working; the environment variables you're overriding don't quite match those you use in the code, and some parts of your code use hardcoded variables anyway so they'll ignore them; we should fix that test up so it's entirely interacting with the test board

I've left various thoughts below, but in particular the priorities are:
* Add an assertion to the integration test to check that it's succeeding
* Adjust the end-to-end tests to use your fresh test board
  * We'll need to update app.py to always use the environment variables rather than hardcoded list ids
  * We'll need to adjust the create_board function to set the correct environment variables
  * And we'll need to adjust the xpaths to match; see comments below for a more maintainable pattern for those 
* We should also update the readme to reflect the latest status of things, e.g. python dependencies are pulled in automatically, and tests can be run with `poetry run pytest ...`

File: todo_app/app.py
Comment (I-Georg) [reply todo_app/app.py]: Deleted

File: todo_app/app.py
Comment (I-Georg) [reply todo_app/app.py]: Updated

File: todo_app/e2e_test.py
Comment (I-Georg) [reply todo_app/e2e_test.py]: Thank you, Updated

File: todo_app/e2e_test.py
Comment (I-Georg) [reply todo_app/e2e_test.py]: Updated,thank you

File: todo_app/flask_config.py
Comment (I-Georg) [reply todo_app/flask_config.py:4]: Updated,thank you

File: todo_app/integration_test.py
Comment (I-Georg) [reply todo_app/integration_test.py]: Updated,thank you

File: todo_app/e2e_test.py
Comment (I-Georg) [reply todo_app/e2e_test.py]: Updated,thank you

File: todo_app/e2e_test.py
Comment (I-Georg) [reply todo_app/e2e_test.py:85]: Updated,thank you

File: README.md
Comment (I-Georg) [reply README.md]: Updated,thank you

File: todo_app/.env.test.py
Comment (I-Georg) [reply todo_app/.env.test.py]: Updated,thank you

File: todo_app/integration_test.py
Comment (I-Georg) [reply todo_app/integration_test.py:32]: Updated,thank you

File: .vagrant/machines/default/virtualbox/action_provision
Comment (BenRamchandani) [inline .vagrant/machines/default/virtualbox/action_provision:1]: Can you remove the vagrant files?

If you don't want to actually delete them, just remove them from git:
* Add `.vagrant/` to your `.gitignore` file
* `git rm --cached -r .vagrant/`
* `git commit -m "Delete vagrant files"`

File: .env.test
Comment (BenRamchandani) [inline .env.test:14]: Can you copy these into your `.env.template` file?

File: todo_app/app.py
Comment (BenRamchandani) [inline todo_app/app.py]: I don't think you need `idBoard` here - we never want to change the board a card is on

File: todo_app/app.py
Comment (BenRamchandani) [inline todo_app/app.py]: It would be clearer to inline these "number" variables, i.e.
```suggestion
		for listNumberDone in range(len(jsonResponseDone)):		
```

File: todo_app/app.py
Comment (BenRamchandani) [inline todo_app/app.py]: The idea is to just pass in the view_model as it already contains all the lists of items

File: todo_app/e2e_test.py
Comment (BenRamchandani) [inline todo_app/e2e_test.py:80]: This line fails for me.

If you add some more ids to your html, e.g. for the create item `<form>` element, then you can use that here rather than a fixed offset `p[20]` which can break very easily

File: todo_app/templates/index.html
Comment (BenRamchandani) [inline todo_app/templates/index.html]:
```suggestion
		<h3><b>To do</b></h3>
```
Tags need to be closed in the reverse order they are opened. `<h3>`, `<b>`, `<ul>`, `<table>` and `<tr>` all need matching closing tags.
This is confusing the browser which is then leading to issues in your e2e tests because the page isn't structured how you'd expect it.

A formatter is really helpful here, use the "Format Document" action (alt + shift + F on Windows) in VSCode.

**Review by BenRamchandani — CHANGES_REQUESTED** [review]
Hi, unfortunately your e2e test is failing for me at the moment, can you have a look at it?

I think it would help if you clean up your index.html file so the tags are all closed correctly, that will make it much easier to write working xpath queries!

It would also be good if you can go through the comments and mark them as resolved once they're sorted so we can all see what's left to do.

File: .vagrant/machines/default/virtualbox/action_provision
Comment (I-Georg) [reply .vagrant/machines/default/virtualbox/action_provision:1]: Hi the vagrant files must have commited when I was switching between the modules. Module 4 pr with vagrant has been approved so I am waiting on this one before I merge the code.

File: .env.test
Comment (I-Georg) [reply .env.test:14]: Copied

File: todo_app/app.py
Comment (I-Georg) [reply todo_app/app.py]: Updated,thank you

File: todo_app/app.py
Comment (I-Georg) [reply todo_app/app.py]: Updated,thank you

File: todo_app/app.py
Comment (I-Georg) [reply todo_app/app.py]: Updated, thank you

File: todo_app/templates/index.html
Comment (I-Georg) [reply todo_app/templates/index.html]: Updated,thank you

File: todo_app/e2e_test.py
Comment (I-Georg) [reply todo_app/e2e_test.py:80]: Updated,thank you

File: .vagrant/machines/default/virtualbox/action_provision
Comment (hugh-emerson) [reply .vagrant/machines/default/virtualbox/action_provision:1]: No matter what order you merge things, you will at some point need to run the above commands.

File: README.md
Comment (hugh-emerson) [inline README.md:73]: You shouldn't be asking the reader to install anything like this. The point of pyproject.toml is to list your projects dependencies. Running `poetry install` will install everything listed in there, including `pytest`

`pytest-dotenv` isn't in there, but you don't need it. You are explicitly loading .env.test / .env as required.

File: README.md
Comment (hugh-emerson) [inline README.md:71]: Your md formatting is off.

You can see it with styling in VSCode via a button in the top right or you can see it here:
https://github.com/I-Georg/DevOps-Course-Starter/blob/342ef45111641bfb7242049f3ae67301df3520e4/README.md

**Review by hugh-emerson — APPROVED** [review]
Your tests do seem to be working nicely now 👍 , but your repository could do with some tidying up. See my other comments for suggestions, plus:

- Delete the `.poetry_cache` folder. Not sure how you accidentally created that in this directory.
- Delete geckodriver.log and add it to .gitignore
- You have a blank file called `git`? Delete that
- You have a folder /todo_app/tests_e2e that just contains a log file? Delete that.
- Delete the .env.test.py file. The correct one is called ".env.test"
- Delete the session_items.py file, which is now unused (you just need to delete the unused import in app.py)

It would neater to put your tests in a separate folder from your actual app code.

Merge whenever you're ready

File: README.md
Comment (I-Georg) [reply README.md:73]: .vagrant/machines/default/virtualbox/action_provision  looks the same in vs code

**Pull request merged**
